src/tike/communicators/pool.py

```diff
@@ -125,11 +178,33 @@ def reduce_mean(self, x: list, axis, worker=None) -> cp.array:
         return cp.mean(self.gather(x, worker=worker, axis=axis),
                        keepdims=True, axis=axis)
 
+    def grouped_allreduce(self, x: list, s: int):
+        """All-reduce x by addition within a subset of GPUs."""
+
+        def f(worker, buf, stride):
+            idx = worker // s * s + (worker + stride) % s
+            return cp.add(buf, self._copy_to(x[idx], worker))
+
+        if self.num_workers == 1:
+            return x
+
+        buff = list(x)
+        for stride in range(1, s):
+            buff = self.map(f, self.workers, buff, stride=stride)
+
+        return buff
+
     def map(self, func, *iterables, **kwargs):
         """ThreadPoolExecutor.map, but wraps call in a cuda.Device context."""
 
         def f(worker, *args):
             with cp.cuda.Device(worker):
                 return func(*args, **kwargs)
 
-        return list(super().map(f, self.workers, *iterables))
+        if 'workers' in kwargs:
+            workers = kwargs.get("workers")
+            kwargs.pop("workers")
+        else:
+            workers = self.workers
+
+        return list(super().map(f, workers, *iterables))
```

Comment:
Why do you need to specify a different set of workers from `self.workers`?

Reply:
In that case, changing the function signature to `*iterables, workers=None, **kwargs`, and then set `workers` to `self.workers` if `worker is None`. I think this should work for python 3.6 and later?